ecosystem/indexer-grpc/indexer-grpc-fullnode/src/runtime.rs

```diff
@@ -0,0 +1,192 @@
+            // Make port into a config
+            .serve(address.to_socket_addrs().unwrap().next().unwrap())
+            .await
+            .unwrap();
+        info!(address = address, "[indexer-grpc] Started GRPC server");
+    });
+    Some(runtime)
+}
+
+#[tonic::async_trait]
+impl IndexerStream for IndexerStreamService {
+    type RawDatastreamStream = ResponseStream;
+
+    async fn raw_datastream(
+        &self,
+        req: Request<RawDatastreamRequest>,
+    ) -> Result<Response<Self::RawDatastreamStream>, Status> {
+        let r = req.into_inner();
+        let starting_version = r.starting_version;
+        let processor_task_count = self.processor_task_count;
+        let processor_batch_size = self.processor_batch_size;
+        let output_batch_size = self.output_batch_size;
+
+        let (tx, rx) = mpsc::channel(TRANSACTION_CHANNEL_SIZE);
+        let context = self.context.clone();
+        let mut ma = MovingAverage::new(10_000);
+
+        let ledger_chain_id = context.chain_id().id();
+        tokio::spawn(async move {
+            let mut coordinator = IndexerStreamCoordinator::new(
+                context,
+                starting_version,
+                processor_task_count,
+                processor_batch_size,
+                output_batch_size,
+                tx.clone(),
+            );
+            let init_status =
+                Self::get_status(StatusType::Init, starting_version, None, ledger_chain_id);
+            match tx.send(Result::<_, Status>::Ok(init_status)).await {
+                Ok(_) => {
+                    // TODO: Add request details later
+                    info!("[indexer-grpc] Init connection");
+                },
+                Err(_) => {
+                    panic!("[indexer-grpc] Unable to initialize stream");
+                },
+            }
+            let mut base: u64 = 0;
+            loop {
+                let results = coordinator.process_next_batch().await;
+                let mut is_error = false;
+                let mut max_version = 0;
+                for result in results {
+                    match result {
+                        Ok(end_version) => {
+                            max_version = std::cmp::max(max_version, end_version);
+                        },
+                        Err(e) => {
+                            error!("[indexer-grpc] Error sending to stream: {}", e);
+                            is_error = true;
+                            break;
+                        },
+                    }
+                }
+                if is_error {
+                    break;
+                }
+                let batch_end_status = Self::get_status(
+                    StatusType::BatchEnd,
+                    coordinator.current_version,
+                    Some(max_version),
+                    ledger_chain_id,
+                );
+                match tx.send(Result::<_, Status>::Ok(batch_end_status)).await {
```

Comment:
so earlier, we panic, now we rror log?

Reply:
OK yea we should probably just panic everywhere. I don't want to make @larry-aptos rebase everything so let's just standardize in the next PR.

Reply:
The panic and error + break behavior is relatively similar (the only difference being whether the node restarts) so the current implementation won't break anything at least.